typed MutableBehavior API

patriknw opened the pull request:
I have been pondering the API for mutable actors with Akka Typed, and this is what I think we should do. The PR is not ready for merge, first I want to tell the story about how I ended up here. Read the comments in IntroSpec.java...

patriknw committed:
API suggestion of MutableBehavior

patriknw commented on akka-docs/rst/java/code/jdocs/akka/typed/IntroSpec.java:
```diff
@@ -0,0 +1,229 @@
+      return chatRoom(new ArrayList<ActorRef<SessionEvent>>());
+    }
+
+    private static Behavior<Command> chatRoom(List<ActorRef<SessionEvent>> sessions) {
```
This style is all good (apart from that immutable collections are not commonly used in Java)

patriknw commented on akka-docs/rst/java/code/jdocs/akka/typed/IntroSpec.java:
```diff
@@ -0,0 +1,229 @@
+    }
+    //#chatroom-behavior
+
+    public static Behavior<Command> mutableChatRoom() {
```
This is the "old" API for writing an Actor with mutable state, several issues...

patriknw commented on akka-docs/rst/java/code/jdocs/akka/typed/IntroSpec.java:
```diff
@@ -0,0 +1,229 @@
+
+    public static Behavior<Command> mutableChatRoom() {
+      return Actor.deferred(c -> {
+        final List<ActorRef<SessionEvent>> sessions = new ArrayList<ActorRef<SessionEvent>>();
```
to me it feels alien to not have a class to put the state variables in

ktoso replied:
Yeah, it could also make debugging and profiling hard, which could put off some people.

patriknw commented on akka-docs/rst/java/code/jdocs/akka/typed/IntroSpec.java:
```diff
@@ -0,0 +1,229 @@
+    public static Behavior<Command> mutableChatRoom() {
+      return Actor.deferred(c -> {
+        final List<ActorRef<SessionEvent>> sessions = new ArrayList<ActorRef<SessionEvent>>();
+        int postCount = 0;
```
`immutable var` doesn't work from lambdas

ktoso replied:
Yes, when closing over stuff in java with lambdas the things are "effectively final" - so even without the final word they act as if they would be..

patriknw commented on akka-docs/rst/java/code/jdocs/akka/typed/IntroSpec.java:
```diff
@@ -0,0 +1,229 @@
+        final List<ActorRef<SessionEvent>> sessions = new ArrayList<ActorRef<SessionEvent>>();
+        int postCount = 0;
+
+        return Actor.stateful((ctx, msg) -> {
```
to structure larger actors the message matches should delegate to methods, but there is no place to write such methods without an enclosing class

johanandren replied:
I agree, and it just struck me that may be a valid point on the Scala side as well. Think complex logic with mutable state like in the sharding actors for example.

patriknw commented on akka-docs/rst/java/code/jdocs/akka/typed/IntroSpec.java:
```diff
@@ -0,0 +1,229 @@
+      return Actor.deferred(ctx -> new MutableChatRoom2());
+    }
+
+    public static class MutableChatRoom2 extends ExtensibleBehavior<Command> {
```
this is just a very clunky way of doing it with the "old" API

patriknw commented on akka-docs/rst/java/code/jdocs/akka/typed/IntroSpec.java:
```diff
@@ -0,0 +1,229 @@
+      }
+    }
+
+    public static Behavior<Command> mutableChatRoom3() {
```
Below is my API suggestion.
Yeah, it has many similarities to current untyped Actor API, and I think that is another advantage. Migration will be easier.

I have used Java to illustrate this, but same thing makes sense in Scala.

ktoso replied:
I like this a lot. Several up-sides from what I can see. Perhaps some of them are too trivial to even mention, but let's do so anyway:

- looks much like old Actors, but now has types. Very likely many users will migrate to this first, and we avoid the risk of being too super different with the new API scaring away people (both new and old). Seems like refactoring many untyped actors to this should be pretty simple, good.
- AFAICS it seems such mutable behaviour could be trivially reused in a fully functional otherwise Actor.
- Very easy to become a mutable behaviour from a stateful/stateless one: `return Actor.stateful((ctx, command) -> { /*...*/ return new MutableChatRoom3(ctx); });` so +1 for reusability

Also, see this draft which makes it possible to write `Actor.mutable(ctx -> msg -> Actor.same())` (thought then it would be more like Actor.create...): https://github.com/ktoso/akka/commit/bd2e89408823c3f3bf3e8f15045833d6075503d5#commitcomment-21797530 This is more for inspiration how these things can be pulled off with target types rather than direclty for the PR.

patriknw commented on akka-docs/rst/java/code/jdocs/akka/typed/IntroSpec.java:
```diff
@@ -0,0 +1,229 @@
+
+      @Override
+      public Behavior<Command> onMessage(Command msg) {
+        if (msg instanceof GetSession) {
```
ignore these ugly `instanceof` for now, we should provide same thing as we have for `AbstractActor`

ktoso replied:
Yeah, the same style as AbstractActor's new Receive will work well here :)

ktoso reviewed (approved):
I played around with this API, I like it a lot.

It will be simple to migrate to from current Actors, and is nicely reusable within "pure" actors and testing too - just constructing with passing in a StatefulContext I gues.

LGTM. Further improvements and ideas may come :) I won't be pushing too crazy things, but would like to consider some of the target typed ideas.

rkuhn commented on akka-docs/rst/scala/code/docs/akka/typed/IntroSpec.scala:
```diff
@@ -66,6 +66,28 @@ object IntroSpec {
+    def mutableChatRoom(): Behavior[Command] =
+      Deferred[Command] { _ =>
+        var sessions: List[ActorRef[SessionEvent]] = List.empty
+        Stateful[Command] { (ctx, msg) ⇒
```
Hmm, this should actually be Stateless, but of course I see the irony in that. (same applies to Java version above)

patriknw replied:
@rkuhn Let me throw some paint on that shed also. As mentioned in #22325 I want to remove `stateless`. Actors are by definition sateful, with some rare exceptions. Symmetry is nice and therefore I suggest that we rename `stateful` to `immutable`. Then we have two (and only two) different styles, and they are so different that they deserve different APIs.

rkuhn commented on akka-typed/src/main/java/akka/typed/javadsl/Actor.java:
```diff
@@ -371,4 +371,28 @@ else if (Behavior.isAlive(behv))
+  }
+
+  //FIXME docs
+  static public abstract class MutableBehavior<T> extends ExtensibleBehavior<T> {
```
Why not directly use ExtensibleBehavior? This class adds very little, which makes me doubt whether it pulls its weight.

rkuhn reviewed (commented):
I can sympathise with the troubles induced by the special restrictions of Java lambdas, but should we not nudge users towards a more immutable style? ExtensibleBehavior already gives full freedom for class-based state management, all that’s needed is documentation.

patriknw commented:
ActorContext is not the right javadsl/scaladsl type. We want to add more, like in AbstractActor, to guide the usage instead of having to copy paste snippets from documentation.

rkuhn commented:
Right, I overlooked that small but significant difference. One aspect that tingles my trouble detectors is that the DSL types will never be suitable for creating behavior combinators because they don’t have the right type of ActorContext available, but that might just be fine because technically such combinators are advanced and possibly non-idiomatic usage (I still prefer a process DSL).

So what you propose is the addition of a MutableBehavior that is specifically for that use-case, and to leave ExtensibleBehavior in `akka.typed` as the extension point.

Concerning the Scala DSL I don’t think we need to add anything since the language already offers all necessary syntactic features for lambda expressions, meaning that Deferred already covers that case.

patriknw commented:
Good. I still think that same thing should exist for Scala DSL:
* more natural to put state variables in a class
* more familiar migration path from classic actors (we will still promote the immutable style as the "best")
* having consistent java and scala APIs is a good thing (e.g. documentation)

patriknw commented:
and yes, `ExtensibleBehavior` is the power user extension point for writing re-usable behaviors

patriknw committed:
cleanup

* moved IntroSpec to akka-typed-tests for more efficient dev environment
* scaladsl.MutableBehavior

patriknw commented:
I have cleaned up and completed this PR.

Added scaladsl also.

I moved the IntroSpec to akka-typed-tests becuase that makes my development environment much more efficient.

patriknw commented:
Refs #22751

patriknw commented:
I see that I should add the mutable to BehaviorSpec also

patriknw committed:
add mutable to BehaviorSpec

patriknw committed:
return this instead of same

johanandren reviewed (approved):
LGTM

2m commented on akka-typed/src/main/java/akka/typed/javadsl/Actor.java:
```diff
@@ -371,4 +371,74 @@ else if (Behavior.isAlive(behv))
+   */
+  static public <T> Behavior<T> mutable(akka.japi.function.Function<ActorContext<T>, MutableBehavior<T>> producer) {
+    return deferred(ctx -> producer.apply(ctx));
+  }
```
What about having an override of this method with a `Supplier` which would be useful for mutable behaviour constructors that do not use context?

2m reviewed (approved):
LGTM
